src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerCopyFileFromContainer.groovy

```diff
@@ -15,55 +15,178 @@
  */
 package com.bmuschko.gradle.docker.tasks.container
 
+import groovy.io.FileType
+import org.gradle.api.GradleException
+import org.gradle.api.file.FileCopyDetails
+import org.gradle.api.file.RelativePath
 import org.gradle.api.tasks.Input
 import org.gradle.api.tasks.Optional
-import org.gradle.api.tasks.OutputFile
-
-import java.util.zip.GZIPOutputStream
 
 class DockerCopyFileFromContainer extends DockerExistingContainer {
+
+    /**
+     * Path inside container
+     */
     @Input
     String resource
 
+    /**
+     * Path on disk to write resource into. If hostPath does
+     * not exist it will be created relative to what we need it
+     * to be (e.g. regular file or directory). This is consistent
+     * with how 'docker cp' behaves.
+     */
     @Input
-    @OutputFile
-    File destFile
+    @Optional
+    File hostPath = project.projectDir
 
+    /**
+     * Whether to leave file in its compressed state or not.
+     *
+     * Docker CP command hands back a tar stream regardless if we asked
+     * for a regular file or directory. Thus, we can give the caller
+     * back the tar file as is or explode it to some location like
+     * 'docker cp' does.
+     */
     @Input
     @Optional
     Boolean compressed = Boolean.FALSE
 
     @Override
     void runRemoteCommand(dockerClient) {
```

Comment:
Broke things up into 2 methods and added more documentation on what exactly is being done and why. It is complex but written so to avoid copying around potentially large files on the host (using cheap renames instead) as we try to create the directory structure the user requested and would expect if doing the same operation via 'docker cp'. Have a look and let me know what you think.